modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -1553,19 +1559,20 @@ public void close() {
      * @return Cluster node to evalute read-only request.
      */
     protected CompletableFuture<ClusterNode> evaluateReadOnlyRecipientNode(int partId) {
-        RaftGroupService svc = raftGroupServiceByPartitionId.get(partId);
+        TablePartitionId tablePartitionId = new TablePartitionId(tableId, partId);
 
-        return svc.refreshAndGetLeaderWithTerm().handle((res, e) -> {
-            if (e != null) {
-                throw withCause(TransactionException::new, REPLICA_UNAVAILABLE_ERR, e);
-            } else {
-                if (res == null || res.leader() == null) {
-                    throw withCause(TransactionException::new, REPLICA_UNAVAILABLE_ERR, e);
-                } else {
-                    return clusterNodeResolver.apply(res.leader().consistentId());
-                }
-            }
-        });
+        return placementDriver.awaitPrimaryReplica(tablePartitionId, clock.now())
+                .orTimeout(AWAIT_PRIMARY_REPLICA_TIMEOUT, TimeUnit.SECONDS).handle((res, e) -> {
+                    if (e != null) {
+                        throw withCause(TransactionException::new, REPLICA_UNAVAILABLE_ERR, e);
+                    } else {
+                        if (res == null || res.getLeaseholder() == null) {
```

Comment:
Do we need the check whether leaseholder is null for the primary replica lease? 'null' leaseholder is possible only for empty lease. And in the same time you removed `null` check from `enlist` method.

Reply:
`|| res.getLeaseholder() == null` removed.